### ISSUE TITLE
When dragging the resize handles, snap to wide / full widths

### ISSUE THREAD

**Issue by jameskoster** [WordPress/gutenberg#44357]
For blocks that have resize handles for width (e.g. Image), it would be nice if the block 'snapped' to predefined widths like wide and full. Perhaps the different widths could appear as guide lines?

<img width="1520" alt="Grid" src="https://user-images.githubusercontent.com/1204802/203553939-5d689c42-73d5-43b4-87e1-1073ce660444.png">

<img width="7440" alt="Only vertical snap areas, tooltip if you pause" src="https://user-images.githubusercontent.com/1204802/203310364-149d7c3a-b693-4da5-8be2-7e85550700b0.png">

Then as you approach a guide line, the width snaps to fill. If you pause for a moment, or as you near a particular snap point, a tooltip appears with the name of the snap point:

<img width="1520" alt="Snap tooltip" src="https://user-images.githubusercontent.com/1204802/203554065-1dcda71c-7077-4a69-bd86-90dc75406ce8.png">

---

Issue updated Nov. 23.

<details><summary>Initial proposal</summary>

**Second iteration:**

![guides](https://user-images.githubusercontent.com/846565/195097355-44fa4485-ecab-4c6a-affd-4924f8bde438.gif)

Then as you approach a guide line, the width snaps to fill. I appreciate this is a bit awkward for wide and full widths as the layout will jump quite a bit, but it's hard to know just how impactful that will be without an interactive prototype (which I cannot make in Figma). Here's the closest approximation I can come up with:

![snapping](https://user-images.githubusercontent.com/846565/195098498-03f9c32d-8ac6-42dd-9be8-c641c3f61a7a.gif)

If the above approach is too technically challenging, a smaller step could be to show snap-zones in a style similar to that of dropzones or the sibling inserter:

<img width="1600" alt="Flow" src="https://user-images.githubusercontent.com/1204802/193565044-d165ca88-d1f1-4a0f-81ab-0edabe2e1cb2.png">

**First iteration:**

<img width="748" alt="Screenshot 2022-09-22 at 10 41 43" src="https://user-images.githubusercontent.com/846565/191714541-1b47e83a-1764-4438-9aa2-752ddd4ae522.png">

For blocks that have resize handles for width (e.g. Image), it would be nice if the block 'snapped' to predefined widths like wide and full.

If the above approach is too technically challenging, a smaller step could be to show snap-zones in a style similar to that of dropzones or the sibling inserter:

<img width="1600" alt="Flow" src="https://user-images.githubusercontent.com/1204802/193565044-d165ca88-d1f1-4a0f-81ab-0edabe2e1cb2.png">

</details>

**Comment by jasmussen** [comment]
Just thinking through what design this needs, here's a GIF showing an image sized up beyond its intrinsic size:

![snapping](https://user-images.githubusercontent.com/1204802/192776938-a26130c0-d458-4787-9e09-d49e1be31f51.gif)

Once you reach a particular wide or full snap-point, what happens? One thought is to show a semi-transparent image overlay at those wide and full positions, and so when releasing the mouse button at those sizes, the alignment is applied.

**Comment by jasmussen** [comment]
Just thinking through what design this needs, here's a GIF showing an image sized up beyond its intrinsic size:

![snapping](https://user-images.githubusercontent.com/1204802/192776938-a26130c0-d458-4787-9e09-d49e1be31f51.gif)

Once you reach a particular wide or full snap-point, what happens? One thought is to show a semi-transparent image overlay at those wide and full positions, and so when releasing the mouse button at those sizes, the alignment is applied.

**Comment by jameskoster** [comment]
I was thinking it would just "stick" to the snap point for _n_ pixels while the mouse is held down. I suppose it's a little tricky because depending on the containers justification the image width could expand both left and right in those positions, unlike the uni-directional scaling in your gif.

**Comment by jasmussen** [comment]
Right, that's exactly the problem, if we actually changed the currently selected block itself, the layout around it can potentially shift and jump because dimensions vastly change. So it may be better to show an overlay which it snaps to on release.

Perhaps the overlay doesn't even need to be a copy of the image, perhaps it's simply a blue indicator of sorts in the width of the wide and full points?

**Comment by jameskoster** [comment]
Yeah that's probably worth a try. Perhaps with some text or icon inclusion to explain what is going on.

**Comment by jasmussen** [comment]
Here's a mockup showing a "snap ruler":

<img width="1600" alt="Flow" src="https://user-images.githubusercontent.com/1204802/193565044-d165ca88-d1f1-4a0f-81ab-0edabe2e1cb2.png">

**Comment by jameskoster** [comment]
Perhaps the different widths could appear as guide lines?

![guides](https://user-images.githubusercontent.com/846565/195097355-44fa4485-ecab-4c6a-affd-4924f8bde438.gif)

Then as you approach a guide line, the width snaps to fill. I appreciate this is a bit awkward for wide and full widths as the layout will jump quite a bit, but it's hard to know just how impactful that will be without an interactive prototype (which I cannot make in Figma). Here's the closest approximation I can come up with:

![snapping](https://user-images.githubusercontent.com/846565/195098498-03f9c32d-8ac6-42dd-9be8-c641c3f61a7a.gif)

**Comment by jasmussen** [comment]
I love that quite a bit. I only wonder how hard that would be to implement, notably the height matching that of the image. But if possible, looks great! I'd use fewer colors, though — perhaps just a single opacity instead of 3? I quite like that it includes "content width" as a snap point, as that would allow you to snap back to responsive if scaling down.

**Comment by jameskoster** [comment]
Re fewer colors, I think that will depend on implementation. In the gif above all overlays have the same opacity, but they are stacked which creates the illusion of multiple colors. 

> I quite like that it includes "content width" as a snap point, as that would allow you to snap back to responsive if scaling down

Yes, I think it can also pave the way for a future where layout is based on a predefined grid. So instead of only seeing content/wide/full guide lines you'd see all the different column widths, and be able to create ad hoc layouts.

**Comment by jasmussen** [comment]
Sounds good. I've updated this issue as marked ready for a dev to tackle it!

**Comment by talldan** [comment]
I'll start taking a look at this one, but open to collaborating with other devs along the way. I think where I'll start is creating the visualizer, and I imagine it to be quite similar to the padding/margin visualizer implementation.

I'd hope it wouldn't then be much of a stretch to also visualize the widths when hovering over the items in the alignment control:
![Screen Shot 2022-10-17 at 4 20 48 pm](https://user-images.githubusercontent.com/677833/196126160-90e7e61a-0970-48b7-aedf-8fb3b4203a29.png)

(but that can be a follow-up to resizing and snapping).

**Comment by talldan** [comment]
An early version of the visualizer in #45056. Feel free to adjust the styles @jameskoster @jasmussen.

There's no snapping yet, that'll be a challenge on its own 😄

**Comment by jasmussen** [comment]
Based on some of the conversations in #45056 I wanted to quickly take a stab at some slightly weaked visuals, to see if there was some simplification to be found. Here's a version with vertical bars only, and the width texts appear as tooltips if you pause for a second or two:
<img width="7440" alt="Only vertical bars, tooltip if you pause" src="https://user-images.githubusercontent.com/1204802/203310314-b821f783-990f-4d9c-8a51-668e359f5373.png">

Here's the same, but with separate column based overlays instead:

<img width="7440" alt="Only vertical snap areas, tooltip if you pause" src="https://user-images.githubusercontent.com/1204802/203310364-149d7c3a-b693-4da5-8be2-7e85550700b0.png">

It shouldn't be that different from what's in progress, but it's slightly simplified. Curiuos what you think, CC: @jameskoster

**Comment by jameskoster** [comment]
Showing the labels situationally seems like a good idea.

Would the first concept work if the lines were full-height, and thinner, so that they appear more like guides? In that design I don't suppose we really need the guides for full width :)

**Comment by jasmussen** [comment]
Maybe, yes. It might also work better for scaling _down_, i.e. "half width" or something, which might then snap to 50% width input with auto height in the dimensions control 🤔

**Comment by jasmussen** [comment]
I updated this issue with some slightly tweaked mockups. They are in part based on the converstion of the PR in progress (#45056), and hopefully they should not be too much of a detour of the code implemented.

**Comment by gracie** [comment]
Sharing feedback from WordPress.com users (via our in-product surveys). Several users asked to resize, crop, and move images by dragging directly on the canvas rather than going through the settings panel — which seemed relevant to this thread on drag resize handles.

> It would be much easier to be able to amend images, like sizes, by using the cursor to reshape things rather than amending the settings, as they are not always easy to find on the right-hand toolbar.

> I miss having more control over the size of pictures with just the move of the mouse.

> I would like to be able to move pictures around more easily.

> Need more options to drag and crop images.

**Comment by ramonjd** [comment]
In relation to the snap widths, the first step could be migrating the image's width block attribute to dimension block supports. 

Image is listed on https://github.com/WordPress/gutenberg/issues/76525 under **"Migrate ad hoc block attributes"**

So a theme could define widths (Core would have sensible defaults):

```json
{
	"style": {
		"blocks": {
			"core/image": {
				"dimensions": {
					"width": true,
					"dimensionSizes": [
						{
							"name": "25%",
							"slug": "25",
							"size": "25%"
						},
						{
							"name": "50%",
							"slug": "50",
							"size": "50%"
						},
						{
							"name": "75%",
							"slug": "75",
							"size": "75%"
						},
						{
							"name": "100%",
							"slug": "100",
							"size": "100%"
						}
					]
				}
			}
		}
	}
}
```

> Allow more flexibility with blocks! Let us drag and drop items, resize them in custom ways, etc
> It would be much easier to be able to amend images, like sizes, by using the cursor to reshape things rather than amending the settings, as they are not always easy to find on the right-hand toolbar.
> I miss having more control over the size of pictures with just the move of the mouse.
> I would like to be able to move pictures around more easily.
> Need more options to drag and crop images.

Hi @gracie 👋🏻 Looking at the issue description, a few things aren't covered:

#### Resizing

Resizing images is already available in the editor.

https://github.com/user-attachments/assets/30dbff7f-0900-4815-8228-19181e11524d

If users are having trouble accessing the resize handles, then I think that'd be a different issue.

#### Block positioning

Moving, dragging and dropping blocks is also a separate issue I think. This one is specifically about defined widths to enable snapping.

Dragging blocks around is possible, albeit in a limited way (between other blocks)

<img width="500" height="260" alt="Image" src="https://github.com/user-attachments/assets/665cc58d-57de-4e63-b48d-cb2ccce098cd" />

<img width="817" height="300" alt="Image" src="https://github.com/user-attachments/assets/5fcfa688-da01-44f1-ab0f-4fbdc51833d7" />

Maybe users want to be able to drag images into existing paragraphs?

There are a few folks thinking about grids and layout in general. See https://github.com/WordPress/gutenberg/issues/81796

#### Cropping

Also separate. The cropping tool is no longer inline, but the tool itself has been replaced by a much richer one with more options. 

<img width="353" height="303" alt="Image" src="https://github.com/user-attachments/assets/e3d56d1d-0185-425c-84fc-5d66824c44c0" />
